include/fmt/core.h

```diff
@@ -1468,13 +1548,38 @@ class dynamic_format_arg_store
   friend class basic_format_args<Context>;
 
   unsigned long long get_types() const {
-    return internal::is_unpacked_bit | data_.size();
+    return internal::is_unpacked_bit | data_.size() |
+           (named_info_.empty() ? 0ULL
+                                : static_cast<unsigned long long>(
+                                      internal::has_named_args_bit));
+  }
+
+  const basic_format_arg<Context>* data() const {
+    return named_info_.empty() ? data_.data() : data_.data() + 1;
   }
 
   template <typename T> void emplace_arg(const T& arg) {
     data_.emplace_back(internal::make_arg<Context>(arg));
   }
 
+  template <typename T>
+  void emplace_arg(const internal::named_arg<T, char_type>& arg) {
+    if (named_info_.empty()) {
+      data_.insert(
+          data_.begin(),
+          {static_cast<const internal::named_arg_info<char_type>*>(nullptr),
+           0});
+    }
+    data_.emplace_back(internal::make_arg<Context>(
+        static_cast<const internal::unwrap_reference_t<T>&>(arg.value)));
+
+    auto guard =
+        internal::make_scope_exit([this]() { this->data_.pop_back(); });
```

Comment:
I suggest using `unique_ptr` with a custom deleter (https://godbolt.org/z/YXFxrR) and not introducing `scope_exit` just for this case.

Reply:
Thanks for pointing -- interesting trick. However to me it looks like a piggyback/abuse of unique_ptr semantic. I realize that it's a trade-off to support more code vs using standard one in an untypical way.
Please could you confirm you prefer unique_ptr here?

Reply:
I would prefer `unique_ptr` or a simple class that stores a pointer to a vector and calls `pop_back` in dtor.

Reply:
Although the class will basically be a reimplementation of a subset of `unique_ptr` with hardcoded deleter. I don't think this is an abuse of `unique_ptr` - it's managing the memory of a single element and we are customizing how is this memory managed which is exactly what deleters are for.

Reply:
Fair enough, will do.